subsys/net/lib/nrf_cloud/coap/src/dtls.c

```diff
@@ -0,0 +1,329 @@
+		return err;
+	}
+
+#if defined(RESTRICT_CIPHERS)
+	LOG_INF("  restrict ciphers");
+	err = setsockopt(sock, SOL_TLS, TLS_CIPHERSUITE_LIST, ciphers, sizeof(ciphers));
+	if (err) {
+		LOG_ERR("Error setting cipherlist: %d", errno);
+		return err;
+	}
+#endif
+#if defined(DUMP_CIPHERLIST)
+	int len;
+	int ciphers[32];
+
+	len = sizeof(ciphers);
+	err = getsockopt(sock, SOL_TLS, TLS_CIPHERSUITE_LIST, ciphers, &len);
+	if (err) {
+		LOG_ERR("Error getting cipherlist: %d", errno);
+	} else {
+		int count = len / sizeof(int);
+
+		LOG_INF("New cipherlist:");
+		for (int i = 0; i < count; i++) {
+			LOG_INF("%d. 0x%04X = %s", i, (unsigned int)ciphers[i],
+			       IS_ENABLED(CONFIG_MBEDTLS) ?
+				mbedtls_ssl_get_ciphersuite_name(ciphers[i]) : "");
+		}
+	}
+#endif
+
+	if (mfw_has_cid) {
+		int cid_option = TLS_DTLS_CID_SUPPORTED;
+
+		LOG_INF("  Enable connection id");
+		err = setsockopt(sock, SOL_TLS, TLS_DTLS_CID, &cid_option, sizeof(cid_option));
+		if (err) {
+			LOG_ERR("Error enabling connection ID: %d", errno);
+		}
+
+		int timeout = TLS_DTLS_HANDSHAKE_TIMEO_123S;
+
+		LOG_INF("  Set handshake timeout %d", timeout);
+		err = setsockopt(sock, SOL_TLS, TLS_DTLS_HANDSHAKE_TIMEO,
+				 &timeout, sizeof(timeout));
+		if (err) {
+			LOG_ERR("Error setting handshake timeout: %d", errno);
+		}
+	}
+
+	enum {
+		NONE = 0,
+		OPTIONAL = 1,
+		REQUIRED = 2,
+	};
+
+	int verify = REQUIRED;
+
+	LOG_INF("  Peer verify: %d", verify);
+	err = setsockopt(sock, SOL_TLS, TLS_PEER_VERIFY, &verify, sizeof(verify));
+	if (err) {
+		LOG_ERR("Failed to setup peer verification, errno %d", errno);
+		return -errno;
+	}
+
+	return err;
+}
+
+bool dtls_cid_is_available(void)
+{
+	return mfw_has_cid;
+}
+
+int dtls_session_save(int sock)
+{
+	int dummy = 0;
+	int err;
+
+	LOG_DBG("  Save DTLS CID session");
+	err = setsockopt(sock, SOL_TLS, TLS_DTLS_CONN_SAVE, &dummy, sizeof(dummy));
+	if (err) {
+		LOG_DBG("Failed to save DTLS CID session, errno %d", errno);
+	}
+	return err;
+}
+
+int dtls_session_load(int sock)
+{
+	int dummy = 0;
+	int err;
+
+	LOG_DBG("  Load DTLS CID session");
+	err = setsockopt(sock, SOL_TLS, TLS_DTLS_CONN_LOAD, &dummy, sizeof(dummy));
+	if (err) {
+		LOG_DBG("Failed to load DTLS CID session, errno %d", errno);
+	}
+	return err;
+}
+
+bool dtls_cid_is_active(int sock)
+{
+	int err = 0;
+
+	if (dtls_cid_active) {
+		return true;
+	}
+
+	int status = 0;
+	int len = sizeof(status);
+
+	if (mfw_ge_version_2) {
+		err = getsockopt(sock, SOL_TLS, TLS_DTLS_HANDSHAKE_STATUS, &status, &len);
+		if (!err) {
+			if (len > 0) {
+				if (status == TLS_DTLS_HANDSHAKE_STATUS_FULL) {
+					LOG_INF("Full DTLS handshake performed");
+				} else if (status == TLS_DTLS_HANDSHAKE_STATUS_CACHED) {
+					LOG_INF("Cached DTLS handshake performed");
+				} else {
+					LOG_WRN("Unknown DTLS handshake status: %d", status);
+				}
+			} else {
+				LOG_WRN("No DTLS status provided");
+			}
+		} else {
+			LOG_ERR("Error retrieving handshake status: %d", errno);
+		}
+	}
```

Comment:
It's informational only, so that we can evaluate whether CID is benefitting us or not.

Reply:
I would advocate removal, then, since that would allow you to get rid of the `mfw_ge_version_2` variable altogether and significantly reduce direct reliance on the presence of the modem

Reply:
The modem (and getsockopt call) will respond with an error if the socket option is not supported in the current modem version.

Reply:
I removed the mfw_ge_version_2 flag and the version checking of the modem firmware, and instead rely on getting an `EOPNOTSUPP` error from `setsockopt()` to indicate if a specific feature is available or not.